### ISSUE TITLE
GDAL/OGR + QGIS read-only support

### ISSUE THREAD

**Issue by bdon** [protomaps/PMTiles#177]
# Motivation

Read PMTiles, particularly vector archives, from the cloud or local storage using QGIS.

Currently MBTiles works by "drag and drop" into QGIS, and may work through adding a Vector Tile layer type. We want the same while also supporting `vsicurl` for accessing tiles via HTTP Ranges. 

The priority should be on vector tiles, since raster tiles are well covered by existing formats. We only care about read support for now, because writing vector tilesets is best left to specialized tools like tippecanoe/planetiler. 

[pmtiles.hpp](https://github.com/protomaps/PMTiles/blob/main/cpp/pmtiles.hpp) is the existing implementation of pmtiles v3 utility functions that is copied into the tippecanoe source tree. This is intentionally written to be unopinionated about I/O and compression, as to re-use those already in GDAL. 

## Links

* GDAL [frmts/mbtilesdataset.cpp](https://github.com/OSGeo/gdal/blob/master/frmts/mbtiles/mbtilesdataset.cpp)
* See [GDAL Wiki on Reading COG with GDAL](https://web.archive.org/web/20230328183924/https://trac.osgeo.org/gdal/wiki/CloudOptimizedGeoTIFF)
* [Current PMTiles specification](https://github.com/protomaps/PMTiles/blob/main/spec/v3/spec.md) See #172 for some more details

Tagging @rouault (via @cholmes)

**Comment by rouault** [comment]
OGR PMTiles driver proposed per https://github.com/OSGeo/gdal/pull/7938

**Comment by bdon** [comment]
the GDAL PR addresses opening PMTiles to retrieve OGR features, however given the above requirements maybe what we want is a QGIS native driver analogous to [qgsmbtiles.cpp](https://github.com/qgis/QGIS/blob/653c11f8f899dd73bf5ec0296747714867909c37/src/core/qgsmbtiles.cpp) for PMTiles. This would enable proper dynamic zoom and tile display in QGIS. 

However, the QGIS native driver for MBTiles opens a SQLite database in-process, and can't benefit from the existing abstraction over `/vsicurl/` etc that GDAL provides. The infrastructure for this already exists at [qgsvectortileconnection.cpp](https://github.com/qgis/QGIS/blob/653c11f8f899dd73bf5ec0296747714867909c37/src/core/vectortile/qgsvectortileconnection.cpp)

**Comment by bdon** [comment]
@four43 that is correct, for multi-resolution tiled data that chooses a zoom level with the viewport, the driver needs to be implemented in QGIS itself and not in GDAL (MBTiles for example)

**Comment by wipfli** [comment]
Or maybe even simpler just look at the pmtiles file on pmtiles.io: https://pmtiles.io/#url=https%3A%2F%2Fdata.source.coop%2Fvida%2Fgoogle-microsoft-osm-open-buildings%2Fpmtiles%2Fgoog_msft_osm.pmtiles&map=12.93/47.36621/8.53346

**Comment by jo-chemla** [comment]
Thanks that `pmtiles serve` CLI looks like a good idea. I've tried multiple variations and finally stumbled on the below working one - putting here in case it's useful to anyone, could not find an example with HTTP, only `s3 or file:///` within the docs and repo

Adding the below mvt vector tile layer does indeed loads way faster than the pmtiles url as a vector layer file/service, so thank a lot!

```bash
pmtiles serve . --bucket=https://data.source.coop/vida/google-microsoft-osm-open-buildings/pmtiles --port 8081
# http://127.0.0.1:8081/goog_msft_osm/{z}/{x}/{y}.mvt
```
